src/sync/mod.rs

```diff
@@ -0,0 +1,355 @@
+    pub fn new(
+        keys: impl IntoIterator<Item = (btc::PublicKey, btc::PrivateKey)>,
+        api_client: T,
+    ) -> Self {
+        Self {
+            key_pool: Arc::new(keys.into_iter().collect()),
+            api_client,
+        }
+    }
+
+    /// Perform a one attempt to sign an anchoring proposal, if any.
+    pub fn process(&self) -> Result<(), AnchoringChainUpdateError<T::Error>> {
+        log::trace!("Perform an anchoring chain update");
+
+        match self
+            .api_client
+            .anchoring_proposal()
+            .map_err(AnchoringChainUpdateError::Client)?
+        {
+            AnchoringProposalState::None => Ok(()),
+            AnchoringProposalState::Available {
+                transaction,
+                inputs,
+            } => {
+                let config = self
+                    .api_client
+                    .config()
+                    .map_err(AnchoringChainUpdateError::Client)?;
+                self.handle_proposal(config, transaction, inputs)
+            }
+            AnchoringProposalState::InsufficientFunds { balance, total_fee } => {
+                Err(AnchoringChainUpdateError::InsufficientFunds { balance, total_fee })
+            }
+        }
+    }
+
+    fn handle_proposal(
+        &self,
+        config: Config,
+        proposal: btc::Transaction,
+        inputs: Vec<btc::Transaction>,
+    ) -> Result<(), AnchoringChainUpdateError<T::Error>> {
+        log::trace!("Got an anchoring proposal: {:?}", proposal);
+        // Find among the keys one from which we have a private part.
+        // TODO What we have to do if we find more than one key? [ECR-3222]
+        let keypair = if let Some(keypair) =
+            self.find_private_key(config.anchoring_keys.iter().map(|x| x.bitcoin_key))
+        {
+            keypair
+        } else {
+            return Ok(());
+        };
+        // Create `SignInput` transactions.
+        let redeem_script = config.redeem_script();
+        let block_height = match proposal.anchoring_payload() {
+            Some(payload) => payload.block_height,
+            None => {
+                return Err(AnchoringChainUpdateError::Internal(failure::format_err!(
+                    "Incorrect anchoring proposal found: {:?}",
+                    proposal
+                )))
+            }
+        };
+
+        log::info!(
+            "Found a new unfinished anchoring transaction proposal for height: {}",
+            block_height
+        );
+
+        let mut signer = p2wsh::InputSigner::new(redeem_script);
+        let sign_input_messages = inputs
+            .iter()
+            .enumerate()
+            .map(|(index, proposal_input)| {
+                let signature = signer.sign_input(
+                    TxInRef::new(proposal.as_ref(), index),
+                    inputs[index].as_ref(),
+                    &(keypair.1).0.key,
+                )?;
+
+                signer.verify_input(
```

Comment:
Can this check reasonably fail, or is it more like a sanity check? In other words, would it make sense to move it to a debug assertion?